edx_rest_framework_extensions/auth/jwt/middleware.py

```diff
@@ -258,19 +261,56 @@ def process_view(self, request, view_func, view_args, view_kwargs):  # pylint: d
         # .. custom_attribute_name: has_jwt_cookie
         # .. custom_attribute_description: Enables us to see requests which have the full reconstituted
         #      JWT cookie. If this attribute is missing, it is assumed to be False.
-        monitoring.set_custom_attribute('has_jwt_cookie', has_reconstituted_jwt_cookie)
-
+        set_custom_attribute('has_jwt_cookie', has_reconstituted_jwt_cookie)
+
+        # DRF authentication does not set the request.user early enough for it to be used in process_request/
+        # process_view of middleware. This code enables JWT cookie authentication to set the request.user for
+        # middleware, before it will presumably happen again during DRF authentication.
+        # For more info on DRF and the request.user, see
+        # https://github.com/jpadilla/django-rest-framework-jwt/issues/45#issuecomment-74996698
+        #
+        # If the user has already been authenticated, we already have a user and intentionally avoid resetting it. As of
+        # Oct 2023, this would likely be due to session authentication in middleware. It is thus important for
+        # JwtAuthentication to verify that the session user and JWT user match. It is possible that this would be better
+        # handled through a more traditional AuthenticationMiddleware that handles both JWT cookies and sessions in
+        # the future.
         if has_reconstituted_jwt_cookie and get_setting(ENABLE_SET_REQUEST_USER_FOR_JWT_COOKIE):
-            # DRF does not set request.user until process_response. This makes it available in process_view.
-            # For more info, see https://github.com/jpadilla/django-rest-framework-jwt/issues/45#issuecomment-74996698
-            request.user = SimpleLazyObject(lambda: _get_user_from_jwt(request, view_func))
+            # Since this call to the user is not made lazily, and has the potential to cause issues, we
+            # ensure it is only used in the case of ENABLE_SET_REQUEST_USER_FOR_JWT_COOKIE.
+            if not get_user(request).is_authenticated:
+                # Similar to django/contrib/auth/middleware.py AuthenticationMiddleware.
+                set_flag_is_request_user_set_for_jwt_auth()
+                request.user = SimpleLazyObject(lambda: _get_cached_user_from_jwt(request, view_func))
```

Comment:
**Note to reviewer:** A flag was used, rather than looking for `SimpleLazyObject`, because Django's AuthenticationMiddleware also uses `SimpleLazyObject`. The flag makes things 100% explicit, so I decided on this approach.

@feanil: Feel free to comment or mark as resolved.

Reply:
Cool, that makes sense, I like the explicitness.